File: backend/app/schemas/memory.py
"""Memory center schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class MemoryBase(BaseModel):
    memory_key: str
    name: str
    description: str | None = None
    category: str
    risk_level: str
    version: str
    built_in: bool
    enabled: bool
    can_read: bool
    can_suggest: bool
    can_auto_write: bool
    need_confirm: bool
    audit_log: bool
    related_keys: list[str]
    sort: int
    created_at: datetime
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class MemoryInfo(MemoryBase):
    id: UUID


class MemoryVersionInfo(BaseModel):
    id: UUID
    version: str
    content: str
    change_type: str
    note: str | None = None
    updater: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MemorySuggestionInfo(BaseModel):
    id: UUID
    target_key: str
    text: str
    source: str | None = None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class MemoryDetail(MemoryInfo):
    content: str
    versions: list[MemoryVersionInfo] = []
    suggestions: list[MemorySuggestionInfo] = []


class MemoryCreate(BaseModel):
    memory_key: str = Field(pattern=r"^[a-zA-Z0-9_-]+\.md$")
    name: str | None = None
    description: str | None = None
    content: str | None = None
    category: str | None = None
    risk_level: str | None = None


class MemorySave(BaseModel):
    content: str


class MemoryRollback(BaseModel):
    version: str


class PendingCreate(BaseModel):
    target_key: str = Field(pattern=r"^[a-zA-Z0-9_-]+\.md$")
    text: str
    source: str | None = None


class PendingConfirm(BaseModel):
    text: str | None = None

File: ui/lib/api/memory.ts
import { apiClient } from "./client";
import type { ApiSuccess } from "./auth";

export interface MemoryInfo {
  id: string;
  memory_key: string;
  name: string;
  description: string | null;
  category: string;
  risk_level: string;
  version: string;
  built_in: boolean;
  enabled: boolean;
  can_read: boolean;
  can_suggest: boolean;
  can_auto_write: boolean;
  need_confirm: boolean;
  audit_log: boolean;
  related_keys: string[];
  sort: number;
  created_at: string;
  updated_at: string | null;
}

export interface MemoryVersionInfo {
  id: string;
  version: string;
  content: string;
  change_type: string;
  note: string | null;
  updater: string | null;
  created_at: string;
}

export interface MemorySuggestionInfo {
  id: string;
  target_key: string;
  text: string;
  source: string | null;
  status: string;
  created_at: string;
}

export interface MemoryDetail extends MemoryInfo {
  content: string;
  versions: MemoryVersionInfo[];
  suggestions: MemorySuggestionInfo[];
}

export interface MemoryStats {
  total: number;
  enabled: number;
  high_risk: number;
  pending: number;
  reads: number;
}

export function listMemories() {
  return apiClient.get<ApiSuccess<MemoryInfo[]>>("/memories");
}

export function getMemory(memoryKey: string) {
  return apiClient.get<ApiSuccess<MemoryDetail>>(`/memories/${encodeURIComponent(memoryKey)}`);
}

export function getMemoryStats() {
  return apiClient.get<ApiSuccess<MemoryStats>>("/memories/stats");
}

export function saveMemory(memoryKey: string, content: string) {
  return apiClient.put<ApiSuccess<MemoryInfo>>(`/memories/${encodeURIComponent(memoryKey)}`, { content });
}

export function rollbackMemory(memoryKey: string, version: string) {
  return apiClient.post<ApiSuccess<MemoryInfo>>(
    `/memories/${encodeURIComponent(memoryKey)}/rollback`,
    { version }
  );
}

export function createPendingMemory(targetKey: string, text: string, source?: string) {
  return apiClient.post<ApiSuccess<MemorySuggestionInfo>>("/memories/pending", {
    target_key: targetKey,
    text,
    source,
  });
}

export function confirmPendingMemory(id: string, text?: string) {
  return apiClient.post<ApiSuccess<MemoryInfo>>(`/memories/pending/${id}/confirm`, { text });
}

export function ignorePendingMemory(id: string) {
  return apiClient.post<ApiSuccess<MemorySuggestionInfo>>(`/memories/pending/${id}/ignore`);
}
